src/skywallet/transaction.go

```diff
@@ -0,0 +1,175 @@
+package skywallet
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+
+	messages "github.com/SkycoinProject/hardware-wallet-protob/go"
+
+	"github.com/SkycoinProject/hardware-wallet-go/src/skywallet/wire"
+)
+
+//go:generate mockery -name TransactionSigner -case underscore -inpkg -testonly
+
+// TransactionSigner represents a data (inputs, outputs) needed to sign transaction
+type TransactionSigner interface {
+	Sign() ([]string, error)
+}
+
+// SkycoinTransactionSigner represents signing Skycoin transaction process
+// @used_in TransactionSign
+type SkycoinTransactionSigner struct {
+	Device     *Device
+	Inputs     []*messages.TxAck_TransactionType_TxInputType
+	Outputs    []*messages.TxAck_TransactionType_TxOutputType
+	Version    int
+	LockTime   int
+	signatures []string
+	state      int
+}
+
+// Sign method signs the Skycoin Transaction
+func (s *SkycoinTransactionSigner) Sign() ([]string, error) {
+	msg, err := s.initSigningProcess()
+
+	if err != nil {
+		return nil, err
+	}
+
+	index := 0
+	s.state = 0
+
+	for {
+		switch msg.Kind {
+		case uint16(messages.MessageType_MessageType_TxRequest):
+			txRequest := &messages.TxRequest{}
+			err = proto.Unmarshal(msg.Data, txRequest)
+			if err != nil {
+				return nil, err
+			}
+			switch *txRequest.RequestType {
+			case messages.TxRequest_TXINPUT:
+				if s.state == 0 { // Sending Inputs for InnerHash
+					if len(s.Inputs)-index > 8 {
+						msg, err = s.sendInputs(index, 8)
+						if err != nil {
+							return nil, err
+						}
+						index += 8
+					} else {
+						msg, err = s.sendInputs(index, len(s.Inputs)-index)
+						if err != nil {
+							return nil, err
+						}
+						s.state++
+						index = 0
+					}
+				} else if s.state == 2 { // Sending Inputs for Signatures
+					err = s.addSignatures(&msg)
+					if err != nil {
+						return nil, err
+					}
+					if len(s.Inputs)-index > 8 {
+						msg, err = s.sendInputs(index, 8)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						msg, err = s.sendInputs(index, len(s.Inputs)-index)
+						if err != nil {
+							return nil, err
+						}
+						s.state++
+						index = 0
+					}
+					index += 8
+				} else {
+					return nil, fmt.Errorf("protocol error: unexpected TxRequest type")
+				}
+			case messages.TxRequest_TXOUTPUT:
+				if s.state == 1 { // Sending Outputs for InnerHash
+					if len(s.Outputs)-index > 8 {
+						msg, err = s.sendOutputs(index, 8)
+						if err != nil {
+							return nil, err
+						}
+						index += 8
+					} else {
+						msg, err = s.sendOutputs(index, len(s.Outputs)-index)
+						if err != nil {
+							return nil, err
+						}
+						s.state++
+						index = 0
+					}
+				} else {
+					return nil, fmt.Errorf("protocol error: unexpected TxRequest type")
```

Comment:
TXOUTPUT?

Reply:
Will change to `Expected TxOutput, but got (TxRequest type, which came)`